packages/data/addon/services/navi-dimension.ts

```diff
@@ -22,7 +25,119 @@ export type ServiceOptions = {
   clientId?: string;
 };
 
+type CacheRecord = {
+  time: number; // the last time this record was accessed
+  dimResponse: NaviDimensionResponse; // the actual records
+};
+
 export default class NaviDimensionService extends Service {
+  /**
+   * @property {Object} _dimensionCache - local cache for dimensions
+   */
+  _dimensionCache: Record<string, CacheRecord> = {};
+
+  /**
+   * keeps the cache from getting too large by:
+   *   1) removing items over an hour old from the cache
+   *   2) removing the oldest item in the cache (if cache is over 5 entries)
+   */
+  private _trimCache(): void {
+    const cache = this._dimensionCache;
+    const now = Date.now();
+    let oldestDate = now;
+    let oldestId = '';
+    for (const [id, record] of Object.entries(cache)) {
+      // remove stale items (over an hour old)
+      if (now - record.time > 60 * 60 * 1000) {
+        delete cache[id];
+      }
+      // find oldest non-stale item
+      else if (record.time < oldestDate) {
+        oldestDate = record.time;
+        oldestId = id;
+      }
+    }
+    // if cache still too large, remove oldest item
+    if (Object.keys(cache).length > 5) {
+      delete cache[oldestId];
+    }
+  }
+
+  /**
+   * adds a given dimResponse/cacheId combo to the cache
+   * @param cacheId string (result of the _getCacheId function)
+   * @param dimResponse NaviDimensionResponse
+   */
+  private _addToCache(cacheId: string, dimResponse: NaviDimensionResponse) {
+    if (!cacheId) {
+      return;
+    }
+    const cache = this._dimensionCache;
+    cache[cacheId] = {
+      time: Date.now(),
+      dimResponse: dimResponse,
+    };
+    this._trimCache();
+  }
+
+  /**
+   * fetches a given cacheId's CacheRecord from the _dimensionCache, returns undefined if not in cache
+   * @param cacheId string (result of the _getCacheId function)
+   * @returns CacheRecord | undefined
+   */
+  private _checkCache(cacheId: string): NaviDimensionResponse | undefined {
+    if (!cacheId) {
+      return;
+    }
+    const cache = this._dimensionCache;
+    let cacheResponse = cache[cacheId];
+    let results = undefined;
+
+    // for 'search' type requests, check for cached 'all' requests
+    if (!cacheResponse) {
+      const [dimId, query] = cacheId.split('.');
+      if (query) {
+        cacheResponse = cache[dimId];
+        if (cacheResponse) {
+          // sort the results
+          let searchResults = SearchUtils.searchNaviDimensionRecords(A(cacheResponse.dimResponse.values), query);
+          results = {
+            values: searchResults.map((val) => val.record),
+          };
+        }
+      }
+    } else {
+      results = cacheResponse.dimResponse;
+    }
+
+    // if something was found, update record access time
+    if (cacheResponse) {
+      cacheResponse.time = Date.now();
+    }
+
+    this._trimCache();
```

Comment:
Do we want to do this at the beginning to check for expired cache entries?

Reply:
Depends; moving it up wouldn't change much in practice, but it would mean that the max cache size would be one larger than whatever it's manually set to. We'd also run the risk of removing records and then looking for them in the same service call, but the odds of that edge-case actually happening are kinda low